azure/services/vmextensions/client.go

```diff
@@ -18,138 +18,103 @@ package vmextensions
 
 import (
 	"context"
-	"encoding/json"
 
-	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2021-11-01/compute"
-	"github.com/Azure/go-autorest/autorest"
-	azureautorest "github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute/v5"
 	"github.com/pkg/errors"
-	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
 	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/azure/services/asyncpoller"
 	"sigs.k8s.io/cluster-api-provider-azure/util/reconciler"
 	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
 )
 
 // azureClient contains the Azure go-sdk Client.
 type azureClient struct {
-	vmextensions compute.VirtualMachineExtensionsClient
+	vmextensions *armcompute.VirtualMachineExtensionsClient
 }
 
-// newClient creates a new VM client from subscription ID.
-func newClient(auth azure.Authorizer) *azureClient {
-	c := newVirtualMachineExtensionsClient(auth.SubscriptionID(), auth.BaseURI(), auth.Authorizer())
-	return &azureClient{c}
-}
-
-// newVirtualMachineExtensionsClient creates a new vm extension client from subscription ID.
-func newVirtualMachineExtensionsClient(subscriptionID string, baseURI string, authorizer autorest.Authorizer) compute.VirtualMachineExtensionsClient {
-	vmextensionsClient := compute.NewVirtualMachineExtensionsClientWithBaseURI(baseURI, subscriptionID)
-	azure.SetAutoRestClientDefaults(&vmextensionsClient.Client, authorizer)
-	return vmextensionsClient
+// newClient creates a new vm extensions client from an authorizer.
+func newClient(auth azure.Authorizer) (*azureClient, error) {
+	opts, err := azure.ARMClientOptions(auth.CloudEnvironment())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create virtualmachineextensions client options")
+	}
+	factory, err := armcompute.NewClientFactory(auth.SubscriptionID(), auth.Token(), opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create armcompute client factory")
+	}
+	return &azureClient{factory.NewVirtualMachineExtensionsClient()}, nil
```

Comment:
nit: Is it better to place this logic into a separate helper function like in the  `natgateways` service to keep things consistent? https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3858/files#diff-0e0342ee34be437e8b2443243544f4d3a79fc4655bf9552bffba2f187f4d18ebR46

Reply:
I guess it's a style thing, but I didn't see value in having the `newClient` func separate from the `newNatGatewaysClient` func, so I've made all the new PRs use this style, and I cleaned `natgateways` up to match in #3887.